packages/console-extension/src/index.ts

```diff
@@ -0,0 +1,236 @@
+  activate: (
+    app: JupyterFrontEnd,
+    translator: ITranslator,
+    tracker: IConsoleTracker | null
+  ) => {
+    if (!tracker) {
+      return;
+    }
+
+    const { commands } = app;
+    const trans = translator.load('jupyterlab');
+
+    // wrapper commands to be able to override the icon
+    const runCommand = 'replite:run';
+    commands.addCommand(runCommand, {
+      caption: trans.__('Run'),
+      icon: runIcon,
+      execute: () => {
+        return commands.execute('console:run-forced');
+      }
+    });
+
+    const runButton = new CommandToolbarButton({
+      commands,
+      id: runCommand
+    });
+
+    const restartCommand = 'replite:restart';
+    commands.addCommand(restartCommand, {
+      caption: trans.__('Restart'),
+      icon: refreshIcon,
+      execute: () => {
+        return commands.execute('console:restart-kernel');
+      }
+    });
+
+    const restartButton = new CommandToolbarButton({
+      commands,
+      id: restartCommand
+    });
+
+    const clearCommand = 'replite:clear';
+    commands.addCommand(clearCommand, {
+      caption: trans.__('Clear'),
+      icon: clearIcon,
+      execute: () => {
+        return commands.execute('console:clear');
+      }
+    });
+
+    const clearButton = new CommandToolbarButton({
+      commands,
+      id: clearCommand
+    });
+
+    tracker.widgetAdded.connect((_, console) => {
+      const { toolbar } = console;
+
+      console.toolbar.addItem('run', runButton);
+      console.toolbar.addItem('restart', restartButton);
+      console.toolbar.addItem('clear', clearButton);
+
+      toolbar.addItem('spacer', Toolbar.createSpacerItem());
+
+      const wrapper = new Panel();
+      wrapper.addClass('jp-PoweredBy');
+
+      const node = document.createElement('a');
+      node.textContent = trans.__('Powered by JupyterLite');
+      node.href = 'https://github.com/jupyterlite/jupyterlite';
+      node.target = '_blank';
+      node.rel = 'noopener noreferrer';
+      const poweredBy = new Widget({ node });
+      const icon = new Widget();
+      liteIcon.element({
+        container: icon.node,
+        elementPosition: 'center',
+        margin: '2px 2px 2px 8px',
+        height: 'auto',
+        width: '16px'
+      });
+
+      wrapper.addWidget(poweredBy);
+      wrapper.addWidget(icon);
+      toolbar.addItem('powered-by', wrapper);
+    });
+  }
+};
+
+/**
+ * A plugin to open a code console and
+ * parse custom parameters from the query string arguments.
+ */
+const consolePlugin: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlite/console-extension:console',
+  autoStart: true,
+  optional: [IConsoleTracker, IThemeManager],
+  activate: (
+    app: JupyterFrontEnd,
+    tracker: IConsoleTracker | null,
+    themeManager: IThemeManager | null
+  ) => {
+    if (!tracker) {
+      return;
+    }
+    const { commands } = app;
+
+    const search = window.location.search;
+    const urlParams = new URLSearchParams(search);
```

Comment:
As a follow-on, consider `provide`ing this as a  `IURLParamManager` which can have additional params added that modify how things work at different lifecycle hooks:

```
urlParamManager.registerParam('code', {
  onAppStarted: async () => console.log('whatever')
  onWidgetAdded: (widget) => await ....
});
```

Reply:
Sounds good, adding to the follow-up list.

Maybe this can even be part of `@jupyterlite/application-extension` so it's always available in lite apps.